### DISCUSSION TITLE
GitHub App can't fork repositories

### DISCUSSION THREAD

**Discussion by mikenikles** [community/community#24607] [Apps, API and Webhooks]
<p>Hi :wave:,</p>
<p>According to the REST API docs (<a href="https://docs.github.com/en/rest/repos/forks#create-a-fork" rel="noopener nofollow ugc">link</a>), GitHub Apps should be able to fork a repository.</p>
<p>Unfortunately, that isn’t the case. I experience two different errors, depending on whether I grant the GitHub App access to all repositories or to a selected repository only.</p>
<p><strong>With access to all repositories</strong></p>
<pre><code class="lang-auto">RequestError [HttpError]: You cannot fork this repository to the selected destination due to a policy.
</code></pre>
<p><strong>With access to a selected repository</strong></p>
<pre><code class="lang-auto">RequestError [HttpError]: Resource not accessible by integration
</code></pre>
<p>Below is the code I use:</p>
<pre data-code-wrap="js"><code class="lang-nohighlight">import fs from "fs";
import { App } from "@octokit/app";

const privateKey = fs.readFileSync("./private-key.pem");

const app = new App({
  appId: 12345,
  privateKey,
});

const { data } = await app.octokit.request("/app");
console.log("authenticated as %s", data.name);

for await (const { octokit, repository } of app.eachRepository.iterator()) {
  await octokit.request("POST /repos/{owner}/{repo}/forks", {
    owner: repository.owner.login,
    repo: repository.name,
    organization: "org-id-to-which-the-github-app-belongs"
  })
}
</code></pre>
<p>Does anybody know why this fails?</p>
<p>Thanks,</p>
<p>Mike</p>

**Comment by jsoref** [comment]
<p>I’d file a ticket to <a href="https://support.github.com/">https://support.github.com/</a></p>

**Comment by mikenikles** [comment, marked as answer]
<p>Thanks <a class="mention" href="/jsoref">@jsoref</a> for the tip. I received a response that explains why it doesn’t work in my situation.</p>
<p>For anyone else ending up here:</p>
<blockquote>
<p>“It is currently not possible to fork private repositories due to how GitHub App permission works. GitHub Apps permission is based on installations – each installation’s permissions are separate from the others. And since an installation is scoped to a specific account (the account on which the App was installed), that means that GitHub Apps permissions are limited to a specific account.”</p>
</blockquote>

**Comment by jsoref** [comment]
<p>Makes sense.</p>
<p>For people wondering “what to do?”, you can set up an account, make it a member of the destination org (and if necessary the origin org), and then create a personal access token for it:<br>
<a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token" class="onebox" target="_blank" rel="noopener">https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token</a></p>
<p>And then use that secret instead of the app secret to fork.</p>

**Comment by mikenikles** [comment]
<p>That’s the approach I ended up with – works like a charm. I went for 3 different personal access tokens: dev, preview, prod.</p>

**Comment by rarkins** [comment]
So when the "[Create a Fork](https://docs.github.com/en/rest/repos/forks#create-a-fork)" documentation says "Works with [GitHub Apps](https://docs.github.com/en/developers/apps)", is that a mistake/error and apps cannot create a fork in any way?

**Reply by jsoref** [reply]
Well, it depends on how to define "any". If the app is passed an account token and uses that token, then the app can perform this task.

But, yeah, I'd be tempted to file a bug claiming the indicator is wrong. 

I went and found another one that also seems wrong (transfer repository). But I don't know enough about how these work, so I'm not sure I'd file a bug about either. 

And I ran into some broken markdown in the process, which of I figure out, I might for a bug for ...

**Reply by rarkins** [reply]
Can you clarify what you mean by "account token"? e.g. let's say all repos are public in the following example:

- Public repo github.com/facebook/jest exists, which doesn't have the app installed
- Org github.com/rarkinsforks has the app installed into "All repositories" with write access

Can I generate an app installation token for rarkinsforks which can fork facebook/jest into rarkinsforks/jest?

**Reply by rarkins** [reply]
BTW if you mean "the app is passed a Personal Access Token and uses the PAT to fork", then that's definitely not what I'd regard as "Works with GitHub Apps"

**Reply by jsoref** [reply]
Yes, I mean a PAT. If you're confident this doesn't qualify, then please file a ticket. (It's generated content, so we can't fix it, so you don't have to consider trying to make a PR.)

**Reply by rarkins** [reply]
Thanks for clarifying. I was planning to use a PAT anyway before I saw this post, so I'll go back to that plan. Although I'm an enterprise customer it's not really worth my time to try convincing GitHub of a docs error through their support. At least now anyone finding this thread should have a definitive answer

**Reply by jsoref** [reply]
Filed https://github.com/github/docs/issues/21806
